blueman/main/Tray.py

```diff
@@ -1,25 +1,36 @@
 from importlib import import_module
 import logging
-
-from blueman.Functions import check_single_instance
+import os
+import sys
 from blueman.main.DBusProxies import AppletService
-
-import gi
-gi.require_version('Gtk', '3.0')
-from gi.repository import GLib, Gio
+from gi.repository import Gio
 
 
-class BluemanTray:
+class BluemanTray(Gio.Application):
     def __init__(self):
-        check_single_instance("blueman-tray")
+        super().__init__(application_id="org.blueman.Tray", flags=Gio.ApplicationFlags.FLAGS_NONE)
+        self._active = False
 
-        applet = AppletService()
+    def do_startup(self):
+        Gio.Application.do_startup(self)
+
+        quit_action = Gio.SimpleAction.new("Quit", None)
+        quit_action.connect("activate", self.quit)
+        self.add_action(quit_action)
 
-        main_loop = GLib.MainLoop()
+    def do_activate(self):
+        if self._active:
+            logging.info("Already running, restarting instance")
+            os.execv(sys.argv[0], sys.argv)
 
-        Gio.bus_watch_name(Gio.BusType.SESSION, 'org.blueman.Applet',
-                           Gio.BusNameWatcherFlags.NONE, None, lambda _connection, _name: main_loop.quit())
+        Gio.bus_watch_name(Gio.BusType.SESSION, 'org.blueman.Applet', Gio.BusNameWatcherFlags.NONE,
+                           self._on_name_appeared, self._on_name_vanished)
+        self.hold()
 
+    def _on_name_appeared(self, _connection, name, _owner):
+        logging.debug("Applet started on name %s, showing indicator" % name)
+
+        applet = AppletService()
         indicator_name = applet.GetStatusIconImplementation()
```

Comment:
I'm not so sure if switching the implementation without killing the tray process works. :thinking:

Reply:
I added logic in the Tray what happens if it is launched when it is already active. If active it will call ``os.execv(sys.argv[0], sys.argv)`` and the new implementation is used. In the StatusIcon plugin it just calls ``launch("bluaman-tray")`` on changed implementation.